sdks/python/apache_beam/io/filesystem.py

```diff
@@ -0,0 +1,416 @@
+    """
+    return (isinstance(other, FileMetadata) and
+            self.path == other.path and
+            self.size_in_bytes == other.size_in_bytes)
+
+  def __hash__(self):
+    return hash((self.path, self.size_in_bytes))
+
+  def __ne__(self, other):
+    return not self.__eq__(other)
+
+  def __repr__(self):
+    return 'FileMetadata(%s, %s)' % (self.path, self.size_in_bytes)
+
+
+class abstractstatic(staticmethod):
+  __slots__ = ()
+
+  def __init__(self, function):
+    super(abstractstatic, self).__init__(function)
+    function.__isabstractmethod__ = True
+
+  __isabstractmethod__ = True
+
+
+class FileSystem(object):
+  """A class that defines the functions that can be performed on a filesystem.
+
+  All methods are abstract and they are for file system providers to
+  implement. Clients should use the FileSystemUtil class to interact with
+  the correct file system based on the provided file pattern scheme.
+  """
+  __metaclass__ = abc.ABCMeta
+
+  @staticmethod
+  def _get_compression_type(path, compression_type):
+    if compression_type == CompressionTypes.AUTO:
+      compression_type = CompressionTypes.detect_compression_type(path)
+    elif not CompressionTypes.is_valid_compression_type(compression_type):
+      raise TypeError('compression_type must be CompressionType object but '
+                      'was %s' % type(compression_type))
+    return compression_type
+
+  @abstractstatic
+  def mkdirs(path):
+    """Recursively create directories for the provided path.
+
+    Args:
+      path: string path of the directory structure that should be created
+
+    Raises:
+      IOError if leaf directory already exists.
+    """
+    raise NotImplementedError
+
+  @abstractstatic
+  def match(pattern, limit=None):
+    """Find all matching paths to the pattern provided.
+
+    Args:
+      pattern: string for the file path pattern to match against
+      limit: Maximum number of responses that need to be fetched
+
+    Returns: List of FileMetadata objects that match the provided pattern.
+    """
+    raise NotImplementedError
+
+  @abstractstatic
+  def create(path, mime_type, compression_type):
+    """Returns a write channel for the given file path.
+
+    Args:
+      path: string path of the file object to be written to the system
+      mime_type: MIME type to specify the type on content in the file object
+      compression_type: Type of compression to be used for this object
+    """
+    raise NotImplementedError
+
+  @abstractstatic
+  def open(path, mime_type, compression_type):
+    """Returns a read channel for the given file path.
+
+    Args:
+      path: string path of the file object to be written to the system
+      mime_type: MIME type to specify the type on content in the file object
+      compression_type: Type of compression to be used for this object
+    """
+    raise NotImplementedError
+
+  @abstractstatic
+  def copy(source, destination):
```

Comment:
@peihe Just to clarify do you mean take in multiple sources & destinations as this already does recursive copy.